src/demo.rs

```diff
@@ -1,59 +1,138 @@
+{
+    type Error = super::Error;
+
+    fn get_tx(&self, txid: &Txid) -> Result<Transaction, Self::Error> {
+        self.electrum_client
+            .borrow_mut()
+            .transaction_get(txid)
+            .map_err(|x| x.into())
     }
 
-    fn get_random_utxo(&self) -> Result<OutPoint, Self::Error> {
-        Ok(OutPoint {
-            txid: Txid::from_hex(
-                "0f3fb1116e30963f1dc6631ad0cd7f00e324de7f3348264a1bba539fb4721c5d",
-            )
-            .unwrap(),
-            vout: 0,
-        })
+    fn is_unspent(&self, txout: &OutPoint) -> Result<bool, Self::Error> {
+        let script = &self.get_tx(&txout.txid)?.output[txout.vout as usize].script_pubkey;
+        let unspent_utxos = &self
+            .electrum_client
+            .borrow_mut()
+            .script_list_unspent(&script)?;
+        Ok(unspent_utxos.into_iter().any(|x| x.tx_hash == txout.txid))
     }
 
-    fn broadcast(&self, tx: &Transaction) -> Result<(), Self::Error> {
-        let bytes = serialize(tx);
-        debug!("Broadcasting: {}", bytes.to_hex());
-        Ok(())
+    fn get_random_utxo(
+        &self,
+        txout: &OutPoint,
+        seed: u64,
+    ) -> Result<Option<OutPoint>, Self::Error> {
+        let mut rng: StdRng = SeedableRng::seed_from_u64(seed);
+        if self.utxo_set.borrow().len() == 0 {
+            let mut txid = &txout.txid;
+            let mut tx = self.get_tx(&txid)?.clone();
+            let mut scripts = HashSet::new();
+
+            for _ in 0..rng.gen_range(1, 3) {
+                // We hit a coinbase!
+                if tx.is_coin_base() {
+                    break;
+                }
+
+                txid = &tx.input[rng.gen_range(0, &tx.input.len())]
+                    .previous_output
+                    .txid;
+                tx = self.get_tx(&txid)?.clone();
+            }
+
+            while self.utxo_set.borrow().len() < self.capacity {
+                // We hit a coinbase!
+                if tx.is_coin_base() {
+                    break;
+                }
+
+                txid = &tx.input[0].previous_output.txid;
+                tx = self.get_tx(&txid)?.clone();
+                scripts.extend(
+                    tx.output
+                        .iter()
+                        .map(|x| x.script_pubkey.clone())
+                        .filter(|x| x.is_v0_p2wpkh())
+                        .collect::<Vec<_>>(),
+                );
+
+                for script in &scripts {
+                    let unspent: Vec<_> = self
+                        .electrum_client
+                        .borrow_mut()
+                        .script_list_unspent(&script)?
+                        .into_iter()
+                        .map(|x| OutPoint {
+                            txid: x.tx_hash,
+                            vout: x.tx_pos as u32,
+                        })
+                        .collect();
+                    self.utxo_set.borrow_mut().extend(unspent);
```

Comment:
I'm not sure it's a good idea to take all the unspent outputs of a tx: imagine the case where my utxo is not the only unspent output in its previous transaction. Just by the fact that the other outputs in the same tx weren't selected in the random set I can find it with pretty high accuracy.